### PR TITLE
migration: Support multiple functions

### PR DESCRIPTION
**Pull request by dzhengfy**
There are several enhancements:
- Support to run multiple functions during migration. In the past,
  we only support to run a single function.
- Support to control when the function should be run, before vm is
  paused or after.
- Support to run a function triggered by a given event
- Support to return the function execution results
- Support to check system disk as default for migrated vm

Signed-off-by: Dan Zheng <dzheng@redhat.com>

### CODE REVIEW

File: virttest/migration.py [hunk 1]
```diff
@@ -270,27 +282,159 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
...
+                        end_event = one_func.get('end_event')
+                        if any([begin_event, end_event]):
+                            return True
+                    else:
+                        # It should be a FunctionType
+                        return False
+            elif isinstance(funcs_to_run, types.FunctionType):
+                return False
+
+        def _run_funcs(funcs_to_run, before_pause, virsh_event_session=None):
```
Comment (chunfuwen) [inline virttest/migration.py:340]: The code of _run_funcs is so big, can we separate it into smaller pieces of code block.

File: virttest/migration.py [hunk 2]
```diff
@@ -270,27 +282,159 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
...
+                               ]
+            :return: boolean, True to collect events, otherwise False
+            """
+            if not funcs_to_run:
+                return False
+            if isinstance(funcs_to_run, list):
+                for one_func in funcs_to_run:
+                    if isinstance(one_func, dict):
+                        begin_event = one_func.get('begin_event')
+                        end_event = one_func.get('end_event')
```
Comment (Yingshun) [inline virttest/migration.py:331]: Can we set more specific argument names? for example, func_before_event or func_after_event...

File: virttest/migration.py [hunk 3]
```diff
@@ -270,27 +282,159 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
         :param options: migration options
         :param thread_timeout: time out seconds for the migration thread running
         :param ignore_status: determine if an exception is raised for errors
-        :param func: the function executed during migration thread is running
+        :param func: the function or list of functions executed during migration
```
Comment (Yingshun) [inline virttest/migration.py:285]: How about adding one more argument for these funcs? It's not easy to understand the meaning if it has multiple types I think.

File: virttest/migration.py [hunk 4]
```diff
@@ -270,27 +282,159 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
...
+                                logging.debug("Run function {}".format(one_func))
+                                one_func()
+                        return
+                    else:
+                        logging.debug("Only support to run the function "
+                                      "after guest is paused")
+                elif isinstance(one_func, dict):
+                    logging.debug("Handle function invoking:%s", one_func)
+                    before_vm_pause = 'yes' == one_func.get('before_pause', 'no')
+                    if before_vm_pause != before_pause:
```
Comment (Yingshun) [inline virttest/migration.py:374]: I think we should raise an exception if the VM is paused before executing a func that need to be run earlier...

File: virttest/migration.py (hunk 1 above)
Comment (dzhengfy) [reply virttest/migration.py]: Updated by dividing into two functions.

File: virttest/migration.py (hunk 2 above)
Comment (dzhengfy) [reply virttest/migration.py]: updated

File: virttest/migration.py (hunk 3 above)
Comment (dzhengfy) [reply virttest/migration.py]: updated by adding a new parameter.

File: virttest/migration.py (hunk 4 above)
Comment (dzhengfy) [reply virttest/migration.py]: updated.

**Comment by dzhengfy** [comment]
Rerun:
 (1/1) type_specific.io-github-autotest-libvirt.virsh.live_migration.domjobabort.before_paused.p2p_live_undefinesource.without_postcopy: PASS (268.31 s)
 (1/1) type_specific.io-github-autotest-libvirt.virsh.live_migration.domjobabort.before_paused.p2p_live.without_postcopy: PASS (260.80 s)
 (1/1) type_specific.io-github-autotest-libvirt.virsh.migrate_options_shared.positive_test.set_get_speed_during_migrate.with_postcopy: PASS (235.76 s)
 (1/1) type_specific.io-github-autotest-libvirt.virsh.live_migration.listen_address.all_ipaddrs.ipv4.migrateuri_ipv4_addr.p2p_live.without_postcopy: PASS (316.25 s)
 (1/1) type_specific.io-github-autotest-libvirt.virsh.migrate_options_shared.positive_test.set_get_maxdowntime_before_migrate.without_postcopy: PASS (218.51 s)

File: virttest/migration.py [hunk 5]
```diff
@@ -271,26 +283,187 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
...
+            if isinstance(funcs_to_run, list):
+                for one_func in funcs_to_run:
+                    if isinstance(one_func, dict):
+                        after_event = one_func.get('after_event')
+                        before_event = one_func.get('before_event')
+                        if any([after_event, before_event]):
+                            return True
+                    else:
+                        # It should be a FunctionType
+                        return False
```
Comment (Yingshun) [inline virttest/migration.py:338]: As we discussed, let's raise an exception here. Thanks!

File: virttest/migration.py [hunk 6]
```diff
@@ -271,26 +283,187 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
...
+                                                     uri=srcuri)
+            virsh_event_session.sendline(cmd)
+            logging.debug("Begin to collect domain events...")
+            return virsh_event_session
+
+        def _need_collect_events(funcs_to_run):
+            """
+            Check if there is a need to run a command to collect domain events
+
+            :param funcs_to_run: the functions to be run. It can be a list
```
Comment (Yingshun) [inline virttest/migration.py:315]: Could you please add some comments to explain that the events will be collected if one of items has "after_event" or "before_event"?

File: virttest/migration.py [hunk 7]
```diff
@@ -271,26 +283,187 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
         :param thread_timeout: time out seconds for the migration thread running
         :param ignore_status: determine if an exception is raised for errors
         :param func: the function executed during migration thread is running
+        :param multi_funcs: list of functions executed during migration
+                     thread is running. The func and multi_funcs should not be
+                     provided at same time.
         :param args: dictionary used by func,
                      'func_param' is mandatory if no real func_param, none is
```
[the remark this answers is not in the document]
Comment (dzhengfy) [reply virttest/migration.py]: updated

File: virttest/migration.py (hunk 6 above)
Comment (dzhengfy) [reply virttest/migration.py:328]: updated

File: virttest/migration.py [hunk 8]
```diff
@@ -271,26 +283,187 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
...
+                    if isinstance(one_func, dict):
+                        after_event = one_func.get('after_event')
+                        before_event = one_func.get('before_event')
+                        if any([after_event, before_event]):
+                            return True
+                    else:
+                        # It should be a FunctionType
+                        return False
+            elif isinstance(funcs_to_run, types.FunctionType):
+                return False
```
[the remark this answers is not in the document]
Comment (dzhengfy) [reply virttest/migration.py:355]: updated

File: virttest/migration.py [hunk 9]
```diff
@@ -271,26 +283,187 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
...
+                                 paused on source host, otherwise, False
+            :param virsh_event_session: VirshSession to collect domain events
+            :raises: exceptions.TestError if any test error happens
+            """
+            for one_func in funcs_to_run:
+                if isinstance(one_func, types.FunctionType):
+                    if not before_pause:
+                        _run_simple_func(vm, one_func)
+                    else:
+                        logging.debug("Only support to run the function "
```
[the remark this answers is not in the document]
Comment (dzhengfy) [reply virttest/migration.py:453]: updated

**Comment by dzhengfy** [comment]
@Yingshun , as you suggested, I remove multi_func_param parameter. Please check again.

File: virttest/migration.py [hunk 10]
```diff
@@ -271,26 +283,196 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
         :param thread_timeout: time out seconds for the migration thread running
         :param ignore_status: determine if an exception is raised for errors
         :param func: the function executed during migration thread is running
+        :param multi_funcs: list of functions executed during migration
+                     thread is running. The func and multi_funcs should not be
+                     provided at same time. For example,
+                     multi_funcs = [{"func": <function check_established at 0x7f5833687510>,
+                                     "after_event": "iteration: '1'",
+                                     "before_event": "Suspended Migrated",
+                                     "before_pause": "yes", "func_name": params},
```
Comment (Yingshun) [inline virttest/migration.py:292]: func_name should be func_param, right?

**Review by Yingshun — COMMENTED** [review]
Others are good to me.

File: virttest/migration.py (hunk 5 above)
Comment (dzhengfy) [reply virttest/migration.py]: Updated

File: virttest/migration.py (hunk 10 above)
Comment (dzhengfy) [reply virttest/migration.py]: Updated and thanks for identifying this typo.

File: virttest/migration.py [hunk 11]
```diff
@@ -271,26 +283,196 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
...
-                     'func_param' is mandatory if no real func_param, none is
-                     requested.
+                     'func_param' is mandatory for func parameter.
+                     If no real func_param, none is requested.
                      'shell' is optional, where shell=True(bool) can be used
                      for process.run
-
         """
+
+        def _run_collect_event_cmd():
```
Comment (chunfuwen) [inline virttest/migration.py:303]: @dzhengfy 
It looks like quite number of embedded methods are newly introduced.
Do you think whether it  is necessary to move them into separate python file ,and let migration.py to use by import?

Comment (dzhengfy) [reply virttest/migration.py:303]: @chunfuwen , these embedded functions are closely related to do_migration() function, so I think there is useless for other functions to reuse them. I prefer to keep them here.

**Review by kylazhang — APPROVED** [review]

File: virttest/migration.py (hunk 11 above)
Comment (kylazhang) [reply virttest/migration.py:303]: Currently seems the embedded functions not that common to move to other file, so I think it is ok to just keep them here.

**Comment by kylazhang** [comment]
As chunfuwen's comments have been already resolved so merge it.

**Pull request merged**